Client/Ecosia/UI/NTP/ClimateImpactCounter/SeedProgressManager/UserDefaultsSeedProgressManager.swift

```diff
@@ -0,0 +1,123 @@
+    }
+
+    // Helper method to get the seed threshold for the current level
+    private static func requiredSeedsForLevel(_ level: Int) -> Int {
+        if let seedLevel = seedLevels.first(where: { $0.level == level }) {
+            return seedLevel.requiredSeeds
+        }
+        return seedLevels.last?.requiredSeeds ?? 0  // If the level exceeds defined levels, use the last one
+    }
+
+    // Calculate the inner progress for the current level (0 to 1)
+    static func calculateInnerProgress() -> CGFloat {
+        let totalSeeds = loadTotalSeedsCollected()
+        let currentLevel = loadCurrentLevel()
+        
+        // Get the required seeds for the current level
+        let thresholdForCurrentLevel = requiredSeedsForLevel(currentLevel)
+        
+        // Seeds needed to reach the current level
+        let previousLevelTotal = currentLevel > 1 ? requiredSeedsForLevel(currentLevel - 1) : 0
+        
+        // Inner progress is calculated between the seeds collected in the current level
+        let seedsForCurrentLevel = totalSeeds - previousLevelTotal
+        return CGFloat(seedsForCurrentLevel) / CGFloat(thresholdForCurrentLevel)
+    }
+
+    // Add seeds to the counter and handle level progression
+    static func addSeeds(_ count: Int) {
+        var totalSeeds = loadTotalSeedsCollected()
+        var currentLevel = loadCurrentLevel()
+
+        let previousLevelTotal = currentLevel > 1 ? requiredSeedsForLevel(currentLevel - 1) : 0
+        let thresholdForCurrentLevel = requiredSeedsForLevel(currentLevel)
+
+        totalSeeds += count
+
+        var leveledUp = false
+        // Only level up if the total seeds EXCEED the threshold for the current level
+        if totalSeeds > previousLevelTotal + thresholdForCurrentLevel {
+            if currentLevel < seedLevels.count {
+                currentLevel += 1
+                leveledUp = true
+            }
+        }
+
+        saveProgress(totalSeeds: totalSeeds, currentLevel: currentLevel, lastAppOpenDate: loadLastAppOpenDate())
+
+        // Notify listeners if leveled up
+        if leveledUp {
+            NotificationCenter.default.post(name: levelUpNotification, object: nil)
+        }
+    }
+    
+    // Reset the counter to the initial state
+    static func resetCounter() {
+        saveProgress(totalSeeds: numberOfSeedsAtStart,
+                     currentLevel: 1,
+                     lastAppOpenDate: .now)
+    }
+
+    // Collect a seed once per day
+    static func collectSeed() {
+        let currentDate = Date()
+        let lastOpenDate = loadLastAppOpenDate()
+        let calendar = Calendar.current
+        
+        if calendar.isDateInToday(lastOpenDate) {
+            return // Seed already collected today
+        }
+        
+        // Add 1 seed and save the last open date as today
+        addSeeds(1)
+        saveProgress(totalSeeds: loadTotalSeedsCollected(), currentLevel: loadCurrentLevel(), lastAppOpenDate: currentDate)
```

Comment:
`addSeeds` already saves the progress, do we actually need to save it again here?

Reply:
Sure, I get the separation, it's indeed nice! Even so, `collectDailySeed` always calls `addSeed` which in turn already saves, that's why this line seems duplicated to me as all the same values will have already been just saved 🤔 I agree to keep that inside `addSeed` and just suggested to remove from `collectDailySeed` if that's the case and I'm not misunderstanding.

Reply:
Makes sense, hadn't realised the date aspect sooner, looks great now 🤩